src/main.rs

```diff
@@ -1,3 +1,72 @@
+extern crate pancurses;
+
+use pancurses::Window;
+use pancurses::{initscr, endwin, Input, noecho};
+
 fn main() {
-    println!("Hello, world!");
+    let window = initscr();
+    window.refresh();
+    window.keypad(true);
+    noecho();
+    render_line((0, 0), (40, 10), &window);
+    loop {
+        match window.getch() {
+            Some(Input::Character('q')) => break,
+            _ => ()
+        }
+    }
+    endwin();
+}
+
+trait RenderOutput {
+    fn set_pixel(&self, x: i32, y: i32);
+    fn get_size(&self) -> (i32, i32);
+}
+
+impl RenderOutput for Window {
+    fn set_pixel(&self, x: i32, y: i32) {
+        self.mvaddch(y, x, '#');
+    }
+
+    fn get_size(&self) -> (i32, i32) {
+        (self.get_max_x(), self.get_max_y())
+    }
+}
+
+fn render_line(p0: (i32, i32), p1: (i32, i32), out: &RenderOutput) {
+    match line_orientation(p0, p1) {
+        LineOrientation::HorizontalIsh => {
+            let ((x0, y0), (x1, y1)) =
+                if p0.0 < p1.0 { (p0, p1) } else { (p1, p0) };
+            let a = ((y1 - y0) as f64) / ((x1 - x0) as f64);
+            let mut y = y0 as f64;
+            for x in x0 ..= x1 {
+                out.set_pixel(x, y as i32);
+                y += a;
+            }
+        }
+        LineOrientation::VerticalIsh => {
+            let ((x0, y0), (x1, y1)) =
+                if p0.1 < p1.1 { (p0, p1) } else { (p1, p0) };
+            let a = ((x1 - x0) as f64) / ((y1 - y0) as f64);
```

Comment:
Hmm it's a good question.  It could definitely happen (for instance, if trying to draw a line from `(0, 0)` to `(0, 0)`).

Empirically it seems to work even when that is the case.  _Integer_ division by 0 definitely seems to cause a panic, but floating point division by 0 like this seems to happily result in a `NaN` value.  While that strikes me as a bit ugly, it turns out to be benign since `a` is never used if `y1 - y0` is `0`, since it is only used for every value in the range `y0..=y1`, which would be the empty range in that case.

Reply:
Ah fascinating! I forgot that the bigger difference ends up in the denominator so, as you say, this only happens from like one point to itself - not when drawing a horizontal or vertical line, for instance. And this is a _great_ reason why my idea for calculating slope first is a _terrible_ idea :sweat_smile: